spec/rspec/core/world_spec.rb

```diff
@@ -21,6 +21,33 @@ module RSpec::Core
         world.register(example_group)
         expect(world.example_groups).to include(example_group)
       end
+
+      it "invokes before_register callbacks" do
+        world.before_register do |example_group|
+          examples     = example_group.examples
+          new_examples = []
+          examples.each do |ex|
+            caps = [{:browserName => :firefox}, {:browserName => :chrome}]
+            caps.each_with_index do |cap, index|
+              ex_with_cap = ex.clone
+              new_id      = "#{ex_with_cap.id}_#{index}_#{cap}"
+              ex_with_cap.instance_variable_set(:@id, new_id)
+              new_examples << ex_with_cap
+            end
+          end
+
+          example_group.instance_variable_set(:@examples, new_examples)
```

Comment:
If there's a need for users to mutate some of the state of the example group, we'll want to provide explicit APIs to do so.  Mutating instance variables will never be supported.

That said, for metadata, we already provide an API for mutating metadata:

http://rspec.info/documentation/3.2/rspec-core/RSpec/Core/Configuration.html#define_derived_metadata-instance_method

For the id, I don't think we can support users mutating it.  IDs are meant to be treated as opaque strings that RSpec generates such that future RSpec runs can use them to refer to the same ID.  They are parsed and used in certain ways such that allowing users to mutate them will almost certainly break features that use IDs.

Seeing how you intend to use this callback makes me think that we need to provide a different way to solve your problem.  I need to go read the example you put together in #2078 to understand what exactly you're trying to do.